### PR TITLE
Improve suggestion for missing fmt str in println

### LINKED ISSUE #52347: Nonsense suggestion for `concat!(x)`
The error for `concat!(x)` is:

```
error: expected a literal
 --> src/main.rs:2:13
  |
2 |     concat!(x);
  |             ^
help: you might be missing a string literal to format with
  |
2 |     concat!("{}", x);
  |             ^^^^^^^
```

That suggestion will clearly not work and gets worse when there are more arguments:

```
error: expected a literal
 --> src/main.rs:2:13
  |
2 |     concat!(x, y, z);
  |             ^
help: you might be missing a string literal to format with
  |
2 |     concat!("{}", x, y, z);
  |             ^^^^^^^

error: expected a literal
 --> src/main.rs:2:16
  |
2 |     concat!(x, y, z);
  |                ^
help: you might be missing a string literal to format with
  |
2 |     concat!(x, "{}", y, z);
  |                ^^^^^^^

error: expected a literal
 --> src/main.rs:2:19
  |
2 |     concat!(x, y, z);
  |                   ^
help: you might be missing a string literal to format with
  |
2 |     concat!(x, y, "{}", z);
  |                   ^^^^^^^
```

This suggestion was added by #51614. cc. @csmoe, @estebank

### PR DESCRIPTION
**Pull request by estebank**
Avoid using `concat!(fmt, "\n")` to improve the diagnostics being
emitted when the first `println!()` argument isn't a formatting string
literal.

Fix #52347.

### CODE REVIEW

**Comment by estebank** [comment]
r? @oli-obk

CC @csmoe I tried a couple of approaches, going down the same path that you mentioned in #52347, and the current PR was the least disruptive change I found.

This will need a perf run to see how much of an impact the change to `println!()` actually is :-/

@bors try

**Comment by estebank** [comment]
cc @ollie27

**Comment by kennytm** [comment]
@rust-timer build effb34d8ab90dbb74ba7e229e244dc9993e3ba35

File: src/libstd/macros.rs [hunk 1]
```diff
@@ -155,8 +155,14 @@ macro_rules! print {
 #[stable(feature = "rust1", since = "1.0.0")]
 macro_rules! println {
     () => (print!("\n"));
-    ($fmt:expr) => (print!(concat!($fmt, "\n")));
-    ($fmt:expr, $($arg:tt)*) => (print!(concat!($fmt, "\n"), $($arg)*));
+    ($fmt:expr) => ({
+        print!($fmt);
+        print!("\n");
```
Comment (ollie27) [inline src/libstd/macros.rs]: Using two `print!`s will mean stdout will be unlocked before printing the `\n`. Replacing these two rules with

```rust
($($arg:tt)*) => (print!("{}\n", format_args!($($arg)*)));
```

will avoid that issue.

File: src/libsyntax_ext/concat.rs [hunk 2]
```diff
@@ -51,17 +52,15 @@ pub fn expand_syntax_ext(
...
-                    format!("\"{{}}\", {}", snippet),
-                );
-                err.emit();
+                missing_literal.push(e.span);
             }
         }
     }
+    if missing_literal.len() > 0 {
+        let mut err = cx.struct_span_err(missing_literal, "expected a literal");
+        err.note("only `&str` literals can be passed to `concat!()`");
```
Comment (ollie27) [inline src/libsyntax_ext/concat.rs]: Most literals can be passed to `concat!`, not just `&str`.

Comment (estebank) [reply src/libsyntax_ext/concat.rs]: True, but the wording "only literals..." can be harder to understand to newcomers... What do you think of
```
only literals (like `"foo"`, `42` and `3.14`) can be passed to `concat!()`
```

**Comment by estebank** [comment]
@kennytm can we get another perf run? After the last change I doubt this will introduce any regression.

**Comment by kennytm** [comment]
@rust-timer build ac3b07a934bb67c24ebcb12eeda45649d0440070

**Comment by kennytm** [comment]
Perf is ready. Looks even worse than the previous one.

**Comment by estebank** [comment]
In that case I guess I'll have to come up with something more involved if we want this :-/

**Comment by estebank** [comment]
Cc @dtolnay after the talk in Rust Meetup. Note to self, see if marching quote literals is possible.

**Comment by dtolnay** [comment]
I can see why performance would be impacted. The new rules simply expand to more code and more indirection. Consider `println!("{}", 0)`.

#### Before

```rust
::io::_print(::std::fmt::Arguments::new_v1_formatted(
    &["", "\n"],
    &match (&0,) {
        (arg0,) => [::std::fmt::ArgumentV1::new(arg0, ::std::fmt::Display::fmt)],
    },
    &[::std::fmt::rt::v1::Argument {
        position: ::std::fmt::rt::v1::Position::At(0usize),
        format: ::std::fmt::rt::v1::FormatSpec {
            fill: ' ',
            align: ::std::fmt::rt::v1::Alignment::Unknown,
            flags: 0u32,
            precision: ::std::fmt::rt::v1::Count::Implied,
            width: ::std::fmt::rt::v1::Count::Implied,
        },
    }],
));
```

#### After

```rust
::io::_print(::std::fmt::Arguments::new_v1_formatted(
    &["", "\n"],
    &match (&::std::fmt::Arguments::new_v1_formatted(
        &[""],
        &match (&0,) {
            (arg0,) => [::std::fmt::ArgumentV1::new(arg0, ::std::fmt::Display::fmt)],
        },
        &[::std::fmt::rt::v1::Argument {
            position: ::std::fmt::rt::v1::Position::At(0usize),
            format: ::std::fmt::rt::v1::FormatSpec {
                fill: ' ',
                align: ::std::fmt::rt::v1::Alignment::Unknown,
                flags: 0u32,
                precision: ::std::fmt::rt::v1::Count::Implied,
                width: ::std::fmt::rt::v1::Count::Implied,
            },
        }],
    ),)
    {
        (arg0,) => [::std::fmt::ArgumentV1::new(arg0, ::std::fmt::Display::fmt)],
    },
    &[::std::fmt::rt::v1::Argument {
        position: ::std::fmt::rt::v1::Position::At(0usize),
        format: ::std::fmt::rt::v1::FormatSpec {
            fill: ' ',
            align: ::std::fmt::rt::v1::Alignment::Unknown,
            flags: 0u32,
            precision: ::std::fmt::rt::v1::Count::Implied,
            width: ::std::fmt::rt::v1::Count::Implied,
        },
    }],
));
```

---

I think there are two ways to land this improvement without a performance impact.

- Somehow improve the diagnostic without changing the macro rules in any way. That could mean recognizing the specific case of `concat!($fmt, "\n")` and knowing to trigger the error on `$fmt` rather than on the entire `concat!` invocation as would otherwise happen.

- Change the macro rules to go through a (private or permanently unstable) new builtin `format_argsln` that is like `format_args` with all the same high-quality diagnostics, but takes care of the trailing `\n`. That way no `concat!` is involved that could interfere with diagnostics.

    ```rust
    #[macro_export]
    #[allow_internal_unstable]
    macro_rules! println {
        ($($arg:tt)*) => {
            $crate::io::_print(format_argsln!($($arg)*))
        };
    }
    ```

File: src/libfmt_macros/lib.rs [hunk 3]
```diff
@@ -273,7 +273,11 @@ impl<'a> Parser<'a> {
             }
         } else {
             let msg = format!("expected `{:?}` but string was terminated", c);
-            let pos = self.input.len() + 1; // point at closing `"`
+            // point at closing `"`, unless the last char is `\n` to account for `println`
+            let pos = match self.input.chars().last() {
+                Some('\n') => self.input.len(),
+                _ => self.input.len() + 1,
```
Comment (estebank) [inline src/libfmt_macros/lib.rs]: This is a hack which will make the span to be shifted one to the left when doing `format!("{\n")`, making the span point at the `\` in stead of the `"`, but the improvement in diagnostics for `println` are worth it given how much of a corner case it is (and we can get back to this later to fix it).

File: src/libstd/macros.rs [hunk 4]
```diff
@@ -153,10 +153,17 @@ macro_rules! print {
...
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[allow_internal_unstable]
 macro_rules! println {
     () => (print!("\n"));
-    ($fmt:expr) => (print!(concat!($fmt, "\n")));
-    ($fmt:expr, $($arg:tt)*) => (print!(concat!($fmt, "\n"), $($arg)*));
+    ($($arg:tt)*) => ({
+        #[cfg(not(stage0))] {
+            ($crate::io::_print(format_args_nl!($($arg)*)));
```
Comment (estebank) [inline src/libstd/macros.rs]: This is needed because `format_args_nl` is not available in `stage0`. This can be safely removed after a short period of time.

**Comment by estebank** [comment]
@bors try
@kennytm perf please?

**Comment by kennytm** [comment]
Please fix the build failure first...

**Comment by kennytm** [comment]
@rust-timer build 4f70795f7992cb13b233f4091ae365f6f36a0eb5

**Comment by estebank** [comment]
@oli-obk I think this is ready for final review. @dtolnay thank you for your suggestion, it ended up being easier than I expected (although I would have to put the same work in for `eprintln!()` if we want the same behavior). @kennytm thank you for running the perf. @ollie27 thank you for the preliminary review and suggestions. All, sorry for the spam.

TL;DR:

- [Perf shows](https://perf.rust-lang.org/compare.html?start=878dd0b5e19e086e608351f33bf12e3625425f20&end=4f70795f7992cb13b233f4091ae365f6f36a0eb5) very little change for most cases, with a few isolated _gains_ 🎉
- `eprintln` as no new behavior, the changes made were reverted
- [Fixed a so far unreported error in the handling of string literal parse errors](https://github.com/rust-lang/rust/pull/52394/commits/f4306ffbfcb2280f180e51ce6421e1c6c28dccec), making the machinery account for raw strings and embedded newlines.
- [`formatln!` now uses](https://github.com/rust-lang/rust/pull/52394/files#diff-f7ad4a5ae3364771b775302f47d00063R157) `io::_print` and a [new `format_args_nl!` unstable macro](https://github.com/rust-lang/rust/pull/52394/files#diff-9594b609d939778093a5e26ba75d7059R692) that is literally the same a s`format_args` only that it adds a `\n` at the end of the format string (only if it's not `stage 0`, this check can be removed after a release cycle)
- Because `println` no longer uses `concat!`, #52347 (suggestion meant only for `println!` being emitted for `concat!`) is fixed, and it even suggests a correct literal depending on the arg count (although it is misleading if any are not `Display`):
```
error: format argument must be a string literal
  --> $DIR/bad_hello.rs:14:14
   |
LL |     println!(3, 4);
   |              ^
help: you might be missing a string literal to format with
   |
LL |     println!("{}, {}", 3, 4);
   |              ^^^^^^^^^
```

- Due to the same change as above, now the work done in #50610 also applies to `println!`, giving much nicer output:

```
error: invalid format string: expected `'}'` but string was terminated
  --> $DIR/format-string-error.rs:14:16
   |
LL |     println!("{");
   |                ^ expected `'}'` in format string
   |
   = note: if you intended to print `{`, you can escape it using `{{`
```

vs

```
error: invalid format string: expected `'}'` but string was terminated
  --> $DIR/format-string-error.rs:12:5
   |
LL |     println!("{");
   |     ^^^^^^^^^^^^^^ expected `'}'` in format string
   |
   = note: if you intended to print `{`, you can escape it using `{{`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

- Changed `concat!()` error handling to emit only one diagnostic per incorrect `concat!()`, as well as included a note to explain its intended use:

```
error: expected a literal
  --> $DIR/bad-concat.rs:15:21
   |
LL |     let _ = concat!(x, y, z, "bar");
   |                     ^  ^  ^
   |
   = note: only literals (like `"foo"`, `42` and `3.14`) can be passed to `concat!()`
```

File: src/libfmt_macros/lib.rs [hunk 5]
```diff
@@ -150,12 +152,20 @@ pub struct Parser<'a> {
     pub errors: Vec<ParseError>,
     /// Current position of implicit positional argument pointer
     curarg: usize,
+    /// The style of the string (raw or not), used to position spans correctly
+    style: syntax::ast::StrStyle,
```
Comment (oli-obk) [inline src/libfmt_macros/lib.rs]: If we made this an `Option<usize>` containing the potential offset, we don't need the libsyntax dependency. We don't really care that it's a raw string, we just need some span offset for diagnostics.

**Review by oli-obk — APPROVED** [review]
r=me with libsyntax dep removed if you agree, otherwise just merge ;)

File: src/libstd/macros.rs [hunk 6]
```diff
@@ -399,6 +406,19 @@ pub mod builtin {
...
+    /// and nothing more. It is perma-unstable.
+    ///
+    /// [`format_args`]: ../std/macro.format_args.html
+    #[doc(hidden)]
+    #[unstable(feature = "println_format_args", issue="0")]
+    #[macro_export]
+    macro_rules! format_args_nl {
+        ($fmt:expr) => ({ /* compiler built-in */ });
+        ($fmt:expr, $($args:tt)*) => ({ /* compiler built-in */ });
+    }
```
Comment (ollie27) [inline src/libstd/macros.rs]: There's no point is adding this documentation if it's just going to be `#[doc(hidden)]`. All of this can just be removed.

File: src/libsyntax_ext/format.rs [hunk 7]
```diff
@@ -695,18 +708,37 @@ pub fn expand_preparsed_format_args(ecx: &mut ExtCtxt,
...
-        None => return DummyResult::raw_expr(sp),
+        Ok(mut fmt) if append_newline => {
+            fmt.node.0 = Symbol::intern(&format!("{}\n", fmt.node.0));
+            fmt
+        }
+        Ok(fmt) => fmt,
+        Err(mut err) => {
+            let sugg_fmt = match args.len() {
+                0 => "{}".to_string(),
+                _ => format!("{}{{}}", "{}, ".repeat(args.len())),
```
Comment (ollie27) [inline src/libsyntax_ext/format.rs]: It's a very minor point but I would not suggest adding the commas so it matches the Python behaviour of using single spaces as separators. So suggest things like `"{} {}"`.

**Review by ollie27 — COMMENTED** [review]
If we're going to fix #30143 for `println` we really should do the same for `eprintln` and `writeln` to get consistent behaviour and diagnostics.

I'm pretty sure `format_args_nl` will need some extra code to make sure it's behind a feature flag. See https://github.com/rust-lang/rust/blob/7db82ccd765cbfe55c3d8a2c434bc6f9b986843d/src/libsyntax_ext/concat_idents.rs#L25-L32 for example.

File: src/libstd/macros.rs (hunk 6 above)
Comment (estebank) [reply src/libstd/macros.rs]: Only reason I added it was to make it easier for people trying to understand the `println` implementation before having learned about compiler built-ins.

File: src/libsyntax_ext/format.rs (hunk 7 above)
Comment (estebank) [reply src/libsyntax_ext/format.rs]: Will change.

**Comment by kennytm** [comment]
```
[01:26:48] failures:
[01:26:48]     [ui] ui/matches.rs
[01:26:48]     [ui] ui/print.rs
[01:26:48]     [ui] ui/print_literal.rs
[01:26:48]     [ui] ui/println_empty_string.rs
```
